[PtRun] JSON Storage Improvements

gokcekantarci opened the pull request:
<!-- Enter a brief description/summary of your PR here. What does it fix/what does it change/how was it tested (even manually, if necessary)? -->
## Summary of the Pull Request
Settings/Query JSON files have been changed so that they will not be deleted if there are no changes in PT updates.
<!-- Please review the items on the PR checklist before submitting-->
## PR Checklist

- [ ] **Closes:** #29621 #31269 
- [ ] **Communication:** I've discussed this with core contributors already. If work hasn't been agreed, this work might be rejected
- [ ] **Tests:** Added/updated and all pass
- [ ] **Dev docs:** Added/updated
- [ ] **Localization:** All end user facing strings can be localized
- [ ] **Documentation updated:** If checked, please file a pull request on [our docs repo](https://github.com/MicrosoftDocs/windows-uwp/tree/docs/hub/powertoys) and link it here: #xxx
- [ ] **New binaries:** Added on the required places
   - [ ] [JSON for signing](https://github.com/microsoft/PowerToys/blob/main/.pipelines/ESRPSigning_core.json) for new binaries
   - [ ] [WXS for installer](https://github.com/microsoft/PowerToys/blob/main/installer/PowerToysSetup/Product.wxs) for new binaries and localization folder
   - [ ] [YML for CI pipeline](https://github.com/microsoft/PowerToys/blob/main/.pipelines/ci/templates/build-powertoys-steps.yml) for new test projects
   - [ ] [YML for signed pipeline](https://github.com/microsoft/PowerToys/blob/main/.pipelines/release.yml)

<!-- Provide a more detailed description of the PR, other things fixed or any additional comments/features here -->
## Detailed Description of the Pull Request / Additional comments

- The version file structure was changed to information.
- The new information file has been changed to keep both the PT version and the default version of the JSON file.
- The query and settings files have different formats, so the ExtractFields function was added to extract the relevant fields.
- CheckVersionMismatch function was added to prevent file checking if there is no version change.
- CheckWithInformatonFiletoClear function checks whether there are any changes in these files when a new version updates.
- JsonSerializerOptions.DefaultIgnoreCondition WhenWritingNull changed to Never. This change was made to be able to write to the information file even if the JSON field is empty.
- Since Pinyin and ImageCache files are stored in the ConcurrentDictionary structure and do not have their own data structure, changes were not applied to these two files.
- A bug was found in Pinyin file saving. Fixed.

<!-- Describe how you validated the behavior. Add automated tests wherever possible, but list manual validation steps taken as well -->
## Validation Steps Performed
- Tested by deleting the AppData\Local\Microsoft\PowerToys folder.
- All files were tested by changing the version without making any changes to the fields of the files.
- All files were tested by changing the version by making changes to the fields of the files.
- All files were tested without changing the version by making changes to the fields of the files.
- All files were tested without making any changes to the fields of the files and without changing the version.

htcfreek commented on src/modules/launcher/Wox.Infrastructure/Storage/JsonStorage`1.cs:
```diff
@@ -20,11 +22,13 @@ public class JsonStorage<T>
+        private readonly object _saveLock = new object();
+
         // use property initialization instead of DefaultValueAttribute
         // easier and flexible for default value of object
         private static readonly JsonSerializerOptions _serializerOptions = new JsonSerializerOptions
         {
-            DefaultIgnoreCondition = System.Text.Json.Serialization.JsonIgnoreCondition.WhenWritingNull,
+            DefaultIgnoreCondition = System.Text.Json.Serialization.JsonIgnoreCondition.Never,
```
Is there an ither way to achieve the same result? Changing this might waste space for "null" value data on the disk.

gokcekantarci replied:
When we do not write the null values ​​to the information file, when we compare the file related to the information file, we may see incompatibility and delete it even if there is no change in the version updates. That's why I changed DefaultIgnoreCondition that way. In addition, Null fields are less in number than filled ones and do not take up a large area. Also, as a user, I discovered areas that I had not noticed before. I think it can also provide useful information to those who examine these files in terms of usage.

htcfreek replied:
Regarding the Plugin options it make a difference of half of the used space and it makes the file better to read/view.

Please validate that the plugin options still don't write Null values to json.

gokcekantarci replied:
The behavior of JSON files has been changed back. A new JsonSerializerOptions has been created for information files. All JSON fields are kept and checked only in information files.

jaimecbernardo commented:
Hi @gokcekantarci , 
This works from what I've tested, but I have some questions.
Why do we need to save the dafault state of a settings file on another file? This one is puzzling me.
It also makes no sense what's being saved in the Default Content in my opinoin. Here's the example from QueryHistory_information.json:
```
{
  "Version": "v0.79.0",
  "DefaultContent": "{\r\n  \u0022Query\u0022: null,\r\n  \u0022ExecutedDateTime\u0022: \u00220001-01-01T00:00:00\u0022\r\n}"
}
```
Can you please walk me through the process of saving such a field?

gokcekantarci commented:
> Hi @gokcekantarci , This works from what I've tested, but I have some questions. Why do we need to save the dafault state of a settings file on another file? This one is puzzling me. It also makes no sense what's being saved in the Default Content in my opinoin. Here's the example from QueryHistory_information.json:
> 
> ```
> {
>   "Version": "v0.79.0",
>   "DefaultContent": "{\r\n  \u0022Query\u0022: null,\r\n  \u0022ExecutedDateTime\u0022: \u00220001-01-01T00:00:00\u0022\r\n}"
> }
> ```
> 
> Can you please walk me through the process of saving such a field?

Hi @jaimecbernardo  ,

In this issue https://github.com/microsoft/PowerToys/issues/4424
The file storage structure was such that these files would be reset with each update to prevent a possible mismatch in JSON fields during version updates.

The important part when checking files is the "Key" in the JSON file. We check whether there is a mismatch in key values. The "Value" of the keys have no meaning in the controls, so it does not matter whether they are null or have a different value.

We were already keeping a version.txt file for each file. Instead of txt file, I made it a new JSON file. In this way, it was possible to store the JSON file format of the existing file along with its version.

In the new structure in Commit, this file is used for controls. The primary check is whether there is a new version. If there is no new version, it continues without checking the information file. When a new version arrives, a JSON structure is checked with the structure in the new version and the file's own information file. If there is no difference between the two, the file is kept. If there is a difference, the existing file is deleted and updated as in the new version, and the information file is also updated.

We use 2 different types of structures in JSON file storage. One of them is PluginJsonStorage. The settings of the plugins are kept here. A file check is performed when opening Powertoys. That's why we make the necessary checks at this stage.

The second is WoxJsonStorage. This is different from the settings structure. QueryHistory and UserSelectedRecord use this structure. In this case, the current structure is saved to the cache after a new search is made. Additionally, these files are written to when Powertoys is shutting down. I proceeded without removing this existing structure. Likewise, when a new search is made, the version check is first made. If there is a new version, it is checked with the information file. If there is an incompatibility here, the relevant file is deleted. When the cache is updated and Powertoys is closed, new searches are saved in the new format along with previous searches.

Since Pinyin and ImageCache files are stored in the ConcurrentDictionary structure and do not have their own data structure, changes were not applied to these two files.

I wrote the steps I tested in the Validation Steps Performed section of PR. If you find a different test method, I would be happy if you share it.

stefansjfw commented on src/modules/launcher/PowerLauncher/ViewModel/SettingWindowViewModel.cs:
```diff
@@ -16,6 +18,28 @@ public SettingWindowViewModel()
+            try
+            {
+                PowerToysRunSettings data = JsonSerializer.Deserialize<PowerToysRunSettings>("{}", _storage.GetInformationSerializerOptions());
+                PowerToysRunSettings fields = _storage.ExtractFields(data, string.Empty);
```
All calls of ExtractFields in this PR passes string.Empty as second argument, which ends up just returning data back. Am I missing something here?

gokcekantarci replied:
You are right!

Example 1:
{
  "Items": [
    {
      "Query": "X",
      "ExecutedDateTime": "2023-12-25T15:00:15.2036625+03:00"
    }
  ]
}

Example 2:
{
  "Shell": 2,
  "ReplaceWinR": false,
  "LeaveShellOpen": false,
  "RunAsAdministrator": false,
  "Count": {}
}

We have different types of JSON Storage files. The fields that are important to us start from different places.

In the first example above

"Items": [

The following fields are important to us.

In the second one

The fields after "{" are important for comparison.

The second parameter of the ExtractFields function, "rootElement", indicates where to get data from. In the first example, it is "Items" and in the second, the head of the file is rootElement.

The reason why I used this function for the beginning of the file is so that when a developer examines this code or wants to change it, developer can understand that the beginning of the file is used here.

It can be deleted according to your suggestion.

stefansjfw commented on src/modules/launcher/PowerLauncher/ViewModel/SettingWindowViewModel.cs:
```diff
@@ -16,6 +18,28 @@ public SettingWindowViewModel()
+            // Check information file for version mismatch
+            try
+            {
+                PowerToysRunSettings data = JsonSerializer.Deserialize<PowerToysRunSettings>("{}", _storage.GetInformationSerializerOptions());
```
why deserializing {} ? why not just create instance of PowerToysRunSettings ?

gokcekantarci replied:
We deserialize it in order to use the newly added _informationSerializerOptions.

```
private static readonly JsonSerializerOptions _informationSerializerOptions = new JsonSerializerOptions
        {
            DefaultIgnoreCondition = System.Text.Json.Serialization.JsonIgnoreCondition.Never,
            IncludeFields = true,
            PropertyNameCaseInsensitive = true,
            WriteIndented = true,
        };
```
Added a different DefaultIgnoreCondition to check all JSON fields.

gokcekantarci commented on src/modules/launcher/Wox.Infrastructure/Storage/JsonStorage`1.cs:
```diff
@@ -122,18 +131,200 @@ private void BackupOriginFile()
+                    JsonElement targetElement = root;
+
+                    // If rootElement is not empty, try to get the specified property
+                    if (!string.IsNullOrEmpty(rootElement) && root.TryGetProperty(rootElement, out var retrievedElement))
```
Done.

crutkas commented:
what is needed to push this forward

gokcekantarci commented:
Comments have been applied.
- ExtractFields function has been removed.
- Object instance is used instead of Deserializing.

Latest main has been merged.

stefansjfw commented on src/modules/launcher/Wox.Infrastructure/Storage/JsonStorage`1.cs:
```diff
@@ -122,18 +126,146 @@ private void BackupOriginFile()
+            }
+        }
+
+        public bool CheckVersionMismatch(T actualData, bool versionMismatch = false)
```
CheckVersionMismatch is called with the default object every time. So this argument is not needed actually I guess.

gokcekantarci replied:
Done.

stefansjfw commented on src/modules/launcher/PowerLauncher/ViewModel/MainViewModel.cs:
```diff
@@ -216,6 +217,48 @@ private void OpenResultsEvent(object index, bool isMouseClick)
+
+                            if (userSelectedRecordItemData != null)
+                            {
+                                if (userSelectedRecordItemStorage.CheckVersionMismatch(userSelectedRecordItemData, _userSelectedRecordStorage.GetVersionMismatch()))
```
Can there be a situation where userSelectedRecordItemStorage._storageHelper.ClearCache has different value than _userSelectedRecordStorage._storageHelper.ClearCache. I believe not? In that case why creating a copy of the storage, can we just use the field member storage?

gokcekantarci replied:
Done.

stefansjfw commented on src/modules/launcher/Wox.Infrastructure/Storage/JsonStorage`1.cs:
```diff
@@ -122,18 +126,146 @@ private void BackupOriginFile()
-                string serialized = JsonSerializer.Serialize(_data, _serializerOptions);
-                File.WriteAllText(FilePath, serialized);
-                _storageHelper.Close();
+                var infoFields = JsonSerializer.Deserialize<Dictionary<string, object>>(File.ReadAllText(infoFilePath), _informationSerializerOptions);
```
I can't figure out why this entire logic is needed? i.e. why can we do something like:
```
try deserialize the file as it is
if it fails, it's either malformed or have new fields. in that case override it
```

could you elaborate on why this wouldn't work?

gokcekantarci commented on src/modules/launcher/Wox.Infrastructure/Alphabet.cs:
```diff
@@ -38,8 +37,8 @@ private void InitializePinyinHelpers()
         Stopwatch.Normal("|Wox.Infrastructure.Alphabet.Initialize|Preload pinyin cache", () =>
         {
-            _pinyinStorage = new WoxJsonStorage<Dictionary<string, string[][]>>("Pinyin");
-            SetPinyinCacheAsDictionary(__cache = _pinyinStorage.Load());
```
Done.

stefansjfw commented on src/modules/launcher/Wox.Infrastructure/Storage/JsonStorage`1.cs:
```diff
@@ -44,20 +57,12 @@ public class JsonStorage<T>
         private StoragePowerToysVersionInfo _storageHelper;
 
-        public T Load()
+        private string DefaultFileContent { get; set; }
```
not used. can be deleted

gokcekantarci replied:
Done.

stefansjfw commented on src/modules/launcher/Wox.Infrastructure/Storage/WoxJsonStorage`1.cs:
```diff
@@ -12,6 +12,7 @@ public class WoxJsonStorage<T> : JsonStorage<T>
     {
         private static readonly IFileSystem FileSystem = new FileSystem();
         private static readonly IPath Path = FileSystem.Path;
+        private static string fileNameJSON;
```
same here. can be removed

stefansjfw commented on src/modules/launcher/Wox.Infrastructure/Storage/JsonStorage`1.cs:
```diff
@@ -122,17 +127,63 @@ private void BackupOriginFile()
+            return true;
+        }
+
+        public bool CheckWithInformationFileToClear(T actualData)
```
Method name is outdated now :)

stefansjfw replied:
also, argument is not needed anymore

gokcekantarci replied:
Done.

stefansjfw commented on src/modules/launcher/PowerLauncher/ViewModel/MainViewModel.cs:
```diff
@@ -1033,7 +1034,32 @@ public void Save()
         {
             if (!_saved)
             {
+                var historyItemData = new QueryHistory();
```
not needed anymore

gokcekantarci replied:
Done.

stefansjfw commented on src/modules/launcher/PowerLauncher/ViewModel/SettingWindowViewModel.cs:
```diff
@@ -16,6 +18,26 @@ public SettingWindowViewModel()
+            // Check information file for version mismatch
+            try
+            {
+                PowerToysRunSettings data = new PowerToysRunSettings();
```
same here

gokcekantarci replied:
Done.

gokcekantarci commented on src/modules/launcher/PowerLauncher/ViewModel/MainViewModel.cs:
```diff
@@ -1033,7 +1034,32 @@ public void Save()
+
                 _historyItemsStorage.Save();
+
+                var userSelectedRecordItemData = new UserSelectedRecord();
```
Done.

stefansjfw commented on src/modules/launcher/PowerLauncher/Storage/QueryHistory.cs:
```diff
@@ -41,5 +41,21 @@ public void Add(string query)
+                    item.ExecutedDateTime = DateTime.Now;
+                }
+
+                if (string.IsNullOrEmpty(item.Query))
```
I'm not sure if this makes sense 🤔 I'm fine with overwriting execution time, but if Query is malformed and we couldn't load it properly, overwriting it with "Default query" doesn't make it usable in the future. I'd suggest if Query is null or empty, just remove that history entry

stefansjfw replied:
let's do this - `if Query is null or empty, just remove that specific history entry`

stefansjfw commented on src/modules/launcher/Wox.Infrastructure/Storage/JsonStorage`1.cs:
```diff
@@ -122,17 +125,58 @@ private void BackupOriginFile()
+            return true;
+        }
+
+        public bool CheckFileToUpdateOrClear()
```
I'd rename this function to something like `IsFileDataCompatible` or `TryLoadData`

stefansjfw reviewed (commented):
Overall, this looks way better now. I'm only thinking whether we should try keep what's valid as it is now (i.e. keep the Update() methods), or if there are changes in the file vs structure, it's likely that every item will be affected and there's no much point of updating every single record.

It really depends on what changed with the new version.. But that means we would need to keep maintaining Update() methods.. Which will be a mess. Let's remove Update methods and add `            UnmappedMemberHandling = System.Text.Json.Serialization.JsonUnmappedMemberHandling.Disallow,
` to _deserilaizeOptions

stefansjfw commented:
> Overall, this looks way better now. I'm only thinking whether we should try keep what's valid as it is now (i.e. keep the Update() methods), or if there are changes in the file vs structure, it's likely that every item will be affected and there's no much point of updating every single record.
> 
> It really depends on what changed with the new version.. But that means we would need to keep maintaining Update() methods.. Which will be a mess. Let's remove Update methods and add ` UnmappedMemberHandling = System.Text.Json.Serialization.JsonUnmappedMemberHandling.Disallow,` to _deserilaizeOptions

Actually I remember now what we talked already.. It makes sense because if we add some new fields in the structure (which is most probable change), we'd lose everything if we add `System.Text.Json.Serialization.JsonUnmappedMemberHandling.Disallow` .. 

So, current status makes sense.. Sorry for the confusion. You can just focus on my comments specific to the code :)

jaimecbernardo reviewed (approved):
Some more data clean up when needed and works as expected.
Some more robustness as well. I'm not sure about doing the clear file only on a version mismatch if loading fails. Guess we could do that in general.
Still, let's go with it since it seems to work for this case and we'll get some info.
Good work!

lin-ycv commented:
can i get some clarification between `ISettingsProvider` and `PluginJsonStorage(ISavable)`, both seems to have the ability to save/get settings, what is the loading priority and which one is prefered?
